File: src/modlunky2/mem/state.py
from dataclasses import dataclass
import enum
from typing import FrozenSet, Optional, Tuple

from modlunky2.mem.entities import EntityType, Player

from modlunky2.mem.memrauder.dsl import (
    array,
    pointer,
    struct_field,
    dc_struct,
    sc_int32,
    sc_uint32,
    sc_int8,
    sc_uint8,
)
from modlunky2.mem.memrauder.model import DictMap
from modlunky2.mem.memrauder.spelunky2 import UidEntityMap, uid_entity_map


class RunRecapFlags(enum.IntFlag):
    PACIFIST = 1 << 1 - 1
    VEGAN = 1 << 2 - 1
    VEGETARIAN = 1 << 3 - 1
    PETTY_CRIMINAL = 1 << 4 - 1
    WANTED_CRIMINAL = 1 << 5 - 1
    CRIME_LORD = 1 << 6 - 1
    KING = 1 << 7 - 1
    QUEEN = 1 << 8 - 1
    FOOL = 1 << 9 - 1
    EGGPLANT = 1 << 10 - 1
    NO_GOLD = 1 << 11 - 1
    LIKED_PETS = 1 << 12 - 1
    LOVED_PETS = 1 << 13 - 1
    TOOK_DAMAGE = 1 << 14 - 1
    USED_ANKH = 1 << 15 - 1
    KILLED_KINGU = 1 << 16 - 1
    KILLED_OSIRIS = 1 << 17 - 1
    NORMAL_ENDING = 1 << 18 - 1
    HARD_ENDING = 1 << 19 - 1
    SPECIAL_ENDING = 1 << 20 - 1
    DIED = 1 << 21 - 1


class HudFlags(enum.IntFlag):
    UPBEAT_DWELLING_MUSIC = 1 << 1 - 1
    RUNNING_TUTORIAL_SPEEDRUN = 1 << 3 - 1
    ALLOW_PAUSE = 1 << 20 - 1
    HAVE_CLOVER = 1 << 23 - 1


class QuestFlags(enum.IntFlag):
    MOON_CHALLENGE = 1 << 25 - 1
    STAR_CHALLENGE = 1 << 26 - 1
    SUN_CHALLENGE = 1 << 27 - 1


class PresenceFlags(enum.IntFlag):
    MOON_CHALLENGE = 1 << 9 - 1
    STAR_CHALLENGE = 1 << 10 - 1
    SUN_CHALLENGE = 1 << 11 - 1


class LoadingState(enum.IntEnum):
    NOT_LOADING = 0
    START = 1
    LOADING = 2
    END = 3


class Screen(enum.IntEnum):
    UNKNOWN = -1
    LOGO = 0
    INTRO = 1
    PROLOGUE = 2
    TITLE = 3
    MAIN_MENU = 4
    OPTIONS = 5
    UNKNOWN_1 = 6
    LEADERBOARDS = 7
    SEED_INPUT = 8
    CHARACTER_SELECT = 9
    TEAM_SELECT = 10
    CAMP = 11
    LEVEL = 12
    LEVEL_TRANSITION = 13
    DEATH = 14
    SPACESHIP = 15
    ENDING = 16
    CREDITS = 17
    SCORES = 18
    CONSTELLATION = 19
    RECAP = 20
    ARENA_MENU = 21
    UNKNOWN_2 = 22
    UNKNOWN_3 = 23
    UNKNOWN_4 = 24
    ARENA_INTRO = 25
    ARENA_MATCH = 26
    ARENA_SCORES = 27
    LOADING_ONLINE = 28
    LOBBY = 29


class Theme(enum.IntEnum):
    BEFORE_FIRST_RUN = 0
    DWELLING = 1
    JUNGLE = 2
    VOLCANA = 3
    OLMEC = 4
    TIDE_POOL = 5
    TEMPLE = 6
    ICE_CAVES = 7
    NEO_BABYLON = 8
    SUNKEN_CITY = 9
    COSMIC_OCEAN = 10
    CITY_OF_GOLD = 11
    DUAT = 12
    ABZU = 13
    TIAMAT = 14
    EGGPLANT_WORLD = 15
    HUNDUN = 16
    BASE_CAMP = 17
    ARENA = 18


class WinState(enum.IntEnum):
    UNKNOWN = -1
    NO_WIN = 0
    TIAMAT = 1
    HUNDUN = 2
    COSMIC_OCEAN = 3


class FeedcodeNotFound(Exception):
    """Failed to find feedcode within Spelunky2 memory."""


@dataclass(frozen=True)
class Items:
    players: Tuple[Optional[Player], ...] = struct_field(
        0x08, array(pointer(dc_struct), 4)
    )


@dataclass(frozen=True)
class State:
    screen_last: Screen = struct_field(0x08, sc_int32, default=Screen.LEVEL_TRANSITION)
    screen: Screen = struct_field(0x0C, sc_int32, default=Screen.LEVEL)
    screen_next: Screen = struct_field(0x10, sc_int32, default=Screen.LEVEL_TRANSITION)
    loading: LoadingState = struct_field(
        0x14, sc_int32, default=LoadingState.NOT_LOADING
    )
    quest_flags: QuestFlags = struct_field(0x38, sc_uint32, default=0)
    # The total amount spent at shops and stolen by leprechauns. This is non-positive during the run.
    # If the run ends in a victory, the bonus will be added to this during the score screen.
    money_shop_total: int = struct_field(0x58, sc_int32, default=0)
    world_start: int = struct_field(0x5C, sc_uint8, default=1)
    level_start: int = struct_field(0x5D, sc_uint8, default=1)
    theme_start: Theme = struct_field(0x5E, sc_uint8, default=Theme.DWELLING)
    time_total: int = struct_field(0x64, sc_uint32, default=1)
    world: int = struct_field(0x68, sc_uint8, default=1)
    world_next: int = struct_field(0x69, sc_uint8, default=1)
    level: int = struct_field(0x6A, sc_uint8, default=1)
    level_next: int = struct_field(0x6B, sc_uint8, default=2)
    theme: Theme = struct_field(0x74, sc_uint8, default=Theme.DWELLING)
    theme_next: Theme = struct_field(0x75, sc_uint8, default=Theme.DWELLING)
    win_state: WinState = struct_field(0x76, sc_int8, default=WinState.NO_WIN)
    waddler_storage: FrozenSet[EntityType] = struct_field(
        0x8C, array(sc_uint32, 99), default=frozenset()
    )
    run_recap_flags: RunRecapFlags = struct_field(0xA34, sc_uint32, default=0)
    hud_flags: HudFlags = struct_field(0xA50, sc_uint32, default=0)
    time_level: int = struct_field(0xA44, sc_uint32, default=0)
    presence_flags: PresenceFlags = struct_field(0xA54, sc_uint32, default=0)
    next_entity_uid: int = struct_field(0x12E0, sc_uint32, default=0)
    items: Optional[Items] = struct_field(0x12F0, pointer(dc_struct), default=None)
    instance_id_to_pointer: UidEntityMap = struct_field(
        0x1348, uid_entity_map, default_factory=DictMap
    )

File: crates/ml2_trackers/src/state.rs
//! `State`: the root object every tracker reads from. Only the subset
//! actually consumed by the six trackers is decoded for now; more
//! fields can be added as needed. Offsets are stable across game
//! versions except when Blitworks bumps them, at which point the
//! drift surfaces via either bogus enum reads (BadEnum error) or
//! transcript replay.

use ml2_mem::{MemStruct, Pointer, Spel2Process};

use crate::entity::{EntityType, Items};
use crate::entity_map::UidEntityMap;
use crate::enums::{LoadingState, Screen, Theme, WinState};
use crate::flags::{HudFlags, PresenceFlags, QuestFlags, RunRecapFlags};

/// Distance from the feedcode marker to the start of the `State`
/// object. The marker sits five bytes into a struct that precedes
/// State by 0x5F bytes.
pub const FEEDCODE_TO_STATE_OFFSET: u64 = 0x5F;

/// Level configuration state hanging off the pointer at
/// `State::theme_info`. Currently used only for CO tracker's sub-theme
/// address; other fields can be added in follow-ups.
#[derive(Debug, MemStruct)]
pub struct ThemeInfo {
    #[offset(0x10)]
    pub sub_theme_address: u64,
}

#[derive(Debug, MemStruct)]
pub struct State {
    #[offset(0x08)]
    pub screen_last: Screen,
    #[offset(0x0C)]
    pub screen: Screen,
    #[offset(0x10)]
    pub screen_next: Screen,
    #[offset(0x14)]
    pub loading: LoadingState,
    #[offset(0x38)]
    pub quest_flags: QuestFlags,

    /// Non-positive during a run; total spent at shops + stolen by
    /// leprechauns. Turns positive on victory when the bonus lands.
    #[offset(0x58)]
    pub money_shop_total: i32,

    #[offset(0x5C)]
    pub world_start: u8,
    #[offset(0x5D)]
    pub level_start: u8,
    #[offset(0x5E)]
    pub theme_start: Theme,

    /// In-run elapsed time in game ticks (60 Hz).
    #[offset(0x64)]
    pub time_total: u32,

    #[offset(0x68)]
    pub world: u8,
    #[offset(0x69)]
    pub world_next: u8,
    #[offset(0x6A)]
    pub level: u8,
    #[offset(0x6B)]
    pub level_next: u8,

    /// Pointer to ThemeInfo. Null on menu / camp; non-null once a level
    /// is loaded. Follow with `theme_info.load(process)`.
    #[offset(0x6C)]
    pub theme_info: Pointer<ThemeInfo>,

    #[offset(0x74)]
    pub theme: Theme,
    #[offset(0x75)]
    pub theme_next: Theme,
    #[offset(0x76)]
    pub win_state: WinState,

    /// Total levels traversed this run. Distinct from `level` (in-area
    /// level number); kept 0-based to match the exe.
    #[offset(0x83)]
    pub level_count: u8,

    /// Fixed 99-slot array of EntityType IDs currently deposited in
    /// Waddler's storage room. Zero slots are empty. Wire layout: 99
    /// contiguous u32 entries starting at 0x8C.
    #[offset(0x8C)]
    pub waddler_storage: [EntityType; 99],

    /// Timers all in game ticks. Level tick resets on transition;
    /// last_level snapshots the previous level's total on transition;
    /// tutorial is the standalone tutorial timer.
    #[offset(0xA40)]
    pub time_last_level: u32,
    #[offset(0xA44)]
    pub time_level: u32,
    #[offset(0xA48)]
    pub time_tutorial: u32,

    #[offset(0xA34)]
    pub run_recap_flags: RunRecapFlags,
    #[offset(0xA50)]
    pub hud_flags: HudFlags,
    #[offset(0xA54)]
    pub presence_flags: PresenceFlags,

    /// Next entity UID the game will assign. Comparing consecutive
    /// values yields how many entities spawned since the last tick
    /// (used by RunState to find TP shadows, deployed ropes, ghost
    /// spawn, etc).
    #[offset(0x12E0)]
    pub next_entity_uid: u32,

    /// Pointer to the four-player Items block. Null until a run
    /// starts; walk `items.players[0]` for the primary player.
    #[offset(0x12F0)]
    pub items: Pointer<Items>,

    /// Robin Hood open-addressing hashmap keyed by entity UID. Used
    /// by chains to walk companions + look up items on the fly.
    #[offset(0x1348)]
    pub instance_id_to_pointer: UidEntityMap,

    #[offset(0x13A0)]
    pub time_startup: u32,
}

impl State {
    /// Attach to the running game, scan for the feedcode marker, read
    /// State at `feedcode - 0x5F`. Errors surface the same variants
    /// they do lower down (NotAttached / FeedcodeMissing / Read).
    pub fn read_current(process: &Spel2Process) -> ml2_mem::Result<Self> {
        let feedcode = process.get_feedcode()?;
        let base = feedcode - FEEDCODE_TO_STATE_OFFSET;
        <Self as MemStruct>::read_from(process, base)
    }
}
